src/analysis/ast_dependency_detector.rs

```
@@ -0,0 +1,974 @@
+    ) -> Vec<&QualifiedContractIdentifier> {
+        let mut lookup = BTreeMap::new();
+        let mut reverse_lookup = Vec::new();
+
+        let mut index: usize = 0;
+
+        if dependencies.is_empty() {
+            return vec![];
+        }
+
+        for (contract, _) in dependencies {
+            lookup.insert(contract, index);
+            reverse_lookup.push(contract);
+            index += 1;
+        }
+
+        let mut graph = Graph::new();
+        for (contract, contract_dependencies) in dependencies {
+            let contract_id = lookup.get(contract).unwrap();
+            graph.add_node(*contract_id);
+            for dep in contract_dependencies.iter() {
+                let dep_id = match lookup.get(dep) {
+                    Some(id) => id,
+                    None => {
+                        println!(
+                            "{}: {} depends on {}, but {} is not found",
+                            red!("error"),
+                            contract,
+                            dep,
+                            dep
+                        );
+                        process::exit(1);
+                    }
+                };
+                graph.add_directed_edge(*contract_id, *dep_id);
+            }
+        }
+
+        let mut walker = GraphWalker::new();
+        let sorted_indexes = walker.get_sorted_dependencies(&graph);
+
+        let cyclic_deps = walker.get_cycling_dependencies(&graph, &sorted_indexes);
+        if let Some(deps) = cyclic_deps {
+            let mut contracts = vec![];
+            for index in deps.iter() {
+                let contract = reverse_lookup[*index];
+                contracts.push(contract.name.as_str());
+            }
+            println!(
+                "{}: cycling dependencies: {}",
+                red!("error"),
+                contracts.join(", ")
+            );
+            process::exit(1);
```

When using the repl and clarinet as a library is a desktop app or a server, these process::exit are pretty incovenient.
I'm working on getting rid of these pattern across the board, could we return an error here instead?

---

Addressed with the last commit.